File: spring/src/main/java/cardsproject/domain/players/CraftingRecipe.java
package cardsproject.domain.players;

import jakarta.persistence.*;

@Entity
@Table(name = "crafting_recipes")
public class CraftingRecipe {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private Integer dustCost = 0;
    private Boolean isAvailable = true;

    // @ManyToOne -> Card, onDelete=PROTECT, relatedName=crafting_recipes, via=cards
    @Column(name = "result_card_id")
    private Long resultCardId;

    // M2M: required_cards -> Card, represented via CraftingIngredient entity (through model); no direct field here

    public Long getId() { return id; }
    public void setId(Long id) { this.id = id; }
    public Integer getDustCost() { return dustCost; }
    public void setDustCost(Integer dustCost) { this.dustCost = dustCost; }
    public Boolean getIsAvailable() { return isAvailable; }
    public void setIsAvailable(Boolean isAvailable) { this.isAvailable = isAvailable; }
    public Long getResultCardId() { return resultCardId; }
    public void setResultCardId(Long resultCardId) { this.resultCardId = resultCardId; }

    // ── Business operations ──────────────────────────────────────────
    public Boolean canCraft(Integer playerId) {
        // TODO: implement canCraft
        return null;
    }
    public void executeCraft(Integer playerId) {
        // TODO: implement executeCraft
    }
    @com.fasterxml.jackson.annotation.JsonIgnore
    public void disable() {
        // TODO: implement disable
    }
    @com.fasterxml.jackson.annotation.JsonIgnore
    public void enable() {
        // TODO: implement enable
    }

    // ── Validation rules ─────────────────────────────────────────────
    @jakarta.validation.constraints.AssertTrue(message = "Crafting recipe must have a dust cost greater than zero")
    @com.fasterxml.jackson.annotation.JsonIgnore
    public boolean isDustCostPositiveValid() {
        return (getDustCost() == null || getDustCost() > 0);
    }
}

File: symfony/src/Entity/Players/CraftingRecipe.php
<?php

namespace App\Entity\Players;

use Doctrine\ORM\Mapping as ORM;
use Symfony\Component\Serializer\Annotation\Groups;
use App\Repository\Players\CraftingRecipeRepository;
use Doctrine\Common\Collections\ArrayCollection;
use Doctrine\Common\Collections\Collection;
use App\Entity\Cards\Card;

#[ORM\Entity(repositoryClass: CraftingRecipeRepository::class)]
#[ORM\Table(name: 'crafting_recipe')]
class CraftingRecipe
{
    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\Column]
    #[Groups(['craftingRecipe:read'])]
    private ?int $id = null;

    #[ORM\Column(type: 'integer')]
    #[Groups(['craftingRecipe:read', 'craftingRecipe:write'])]
    private int $dustCost = 0;

    #[ORM\Column(type: 'boolean')]
    #[Groups(['craftingRecipe:read', 'craftingRecipe:write'])]
    private bool $isAvailable = true;

    #[ORM\ManyToOne(targetEntity: Card::class, inversedBy: 'craftingRecipes')]
    #[ORM\JoinColumn(nullable: false, onDelete: 'RESTRICT')]
    private ?Card $resultCard = null;

    #[ORM\OneToMany(mappedBy: 'recipe', targetEntity: CraftingIngredient::class)]
    private Collection $ingredients;

    public function __construct()
    {
        $this->ingredients = new ArrayCollection();
    }

    public function getId(): ?int
    {
        return $this->id;
    }

    public function getDustCost(): int
    {
        return $this->dustCost;
    }

    public function setDustCost(int $dustCost): static
    {
        $this->dustCost = $dustCost;
        return $this;
    }

    public function getIsAvailable(): bool
    {
        return $this->isAvailable;
    }

    public function setIsAvailable(bool $isAvailable): static
    {
        $this->isAvailable = $isAvailable;
        return $this;
    }

    #[Groups(['craftingRecipe:read'])]
    public function getResultCardId(): ?int
    {
        return $this->resultCard?->getId();
    }

    public function getResultCard(): ?Card
    {
        return $this->resultCard;
    }

    public function setResultCard(?Card $resultCard): static
    {
        $this->resultCard = $resultCard;
        return $this;
    }

    public function getIngredients(): Collection
    {
        return $this->ingredients;
    }

    // ── Validation rules ─────────────────────────────────────────────
    #[\Symfony\Component\Validator\Constraints\IsTrue(message: "Crafting recipe must have a dust cost greater than zero")]
    public function isDustCostPositiveValid(): bool
    {
        return ($this->getDustCost() === null || $this->getDustCost() > 0);
    }

    // ── Business operations ──────────────────────────────────────────

    public function canCraft($playerId): mixed
    {
        // TODO: implement can_craft
        return null;
    }

    public function executeCraft($playerId): void
    {
        // TODO: implement execute_craft
    }

    public function disable(): void
    {
        // TODO: implement disable
    }

    public function enable(): void
    {
        // TODO: implement enable
    }

}
